File: backend/app/schemas/grading.py
"""
Grading & Authentication Schemas

Pydantic models for API request/response handling.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List, Dict
from uuid import UUID

from pydantic import BaseModel, Field


# ============================================
# GRADING COMPANY SCHEMAS
# ============================================

class ServiceLevelResponse(BaseModel):
    id: UUID
    name: str
    code: Optional[str] = None
    max_value: Optional[Decimal] = None
    base_fee: Decimal
    estimated_days: Optional[int] = None
    is_active: bool
    
    class Config:
        from_attributes = True


class GradingCompanyResponse(BaseModel):
    id: UUID
    name: str
    code: str
    website: Optional[str] = None
    service_type: str  # grading, authentication, both
    is_active: bool
    
    class Config:
        from_attributes = True


class GradingCompanyWithLevels(GradingCompanyResponse):
    service_levels: List[ServiceLevelResponse] = []


# ============================================
# CARD GRADING SCHEMAS
# ============================================

class CardGradingItemCreate(BaseModel):
    """Create a grading item from inventory."""
    inventory_id: UUID
    checklist_id: Optional[UUID] = None
    declared_value: Decimal = Decimal("0")
    fee_per_card: Optional[Decimal] = None
    was_signed: bool = False


class CardGradingItemResponse(BaseModel):
    id: UUID
    inventory_id: Optional[UUID] = None
    checklist_id: Optional[UUID] = None
    line_number: Optional[int] = None
    declared_value: Decimal
    fee_per_card: Optional[Decimal] = None
    was_signed: bool
    status: str
    grade_value: Optional[Decimal] = None
    auto_grade: Optional[Decimal] = None
    cert_number: Optional[str] = None
    label_type: Optional[str] = None
    notes: Optional[str] = None
    
    # Nested details
    player_name: Optional[str] = None
    card_number: Optional[str] = None
    product_line_name: Optional[str] = None
    
    class Config:
        from_attributes = True


class CardGradingSubmissionCreate(BaseModel):
    """Create a new grading submission."""
    company_id: UUID
    service_level_id: Optional[UUID] = None
    submitter_id: Optional[UUID] = None  # Third-party submitter (NULL = direct)
    date_submitted: date
    items: List[CardGradingItemCreate]
    submission_number: Optional[str] = None
    reference_number: Optional[str] = None
    shipping_to_cost: Decimal = Decimal("0")
    shipping_to_tracking: Optional[str] = None
    insurance_cost: Decimal = Decimal("0")
    notes: Optional[str] = None


class CardGradingSubmissionResponse(BaseModel):
    id: UUID
    company_id: UUID
    service_level_id: Optional[UUID] = None
    submitter_id: Optional[UUID] = None
    submission_number: Optional[str] = None
    reference_number: Optional[str] = None
    date_submitted: date
    date_shipped: Optional[date] = None
    date_received: Optional[date] = None
    date_graded: Optional[date] = None
    date_shipped_back: Optional[date] = None
    date_returned: Optional[date] = None
    status: str
    grading_fee: Decimal
    shipping_to_cost: Decimal
    shipping_to_tracking: Optional[str] = None
    shipping_return_cost: Decimal
    shipping_return_tracking: Optional[str] = None
    insurance_cost: Decimal
    total_cards: int
    cards_graded: int
    total_declared_value: Decimal
    notes: Optional[str] = None
    items: List[CardGradingItemResponse] = []
    
    # Nested company info
    company_name: Optional[str] = None
    company_code: Optional[str] = None
    service_level_name: Optional[str] = None
    submitter_name: Optional[str] = None
    
    class Config:
        from_attributes = True


class CardGradingStatusUpdate(BaseModel):
    """Update submission status/tracking."""
    status: str  # shipped, received, grading, shipped_back, returned
    date_shipped: Optional[date] = None
    date_received: Optional[date] = None
    date_graded: Optional[date] = None
    date_shipped_back: Optional[date] = None
    date_returned: Optional[date] = None
    shipping_to_tracking: Optional[str] = None
    shipping_return_tracking: Optional[str] = None


class CardGradeResult(BaseModel):
    """Result for a single graded card."""
    item_id: UUID
    status: str  # graded, authentic, altered, counterfeit, ungradeable, lost
    grade_value: Optional[Decimal] = None
    auto_grade: Optional[Decimal] = None
    cert_number: Optional[str] = None
    label_type: Optional[str] = None
    notes: Optional[str] = None


class CardGradingResultsSubmit(BaseModel):
    """Submit grading results for items."""
    item_results: List[CardGradeResult]
    date_returned: Optional[date] = None
    shipping_return_cost: Decimal = Decimal("0")


class CardGradingStats(BaseModel):
    """Statistics for card grading."""
    pending_submissions: int
    cards_out_for_grading: int
    pending_fees: Decimal
    total_graded: int
    grade_distribution: Dict[str, int]  # {"10": 5, "9.5": 10, ...}
    gem_rate: float  # Percentage of 10s
    avg_grade: Optional[float] = None
    by_company: Dict[str, int]  # {"PSA": 100, "BGS": 50}


class PendingByCompany(BaseModel):
    """Pending grading items grouped by company."""
    company_id: UUID
    company_name: str
    company_code: str
    pending_count: int
    pending_value: Decimal
    oldest_submission_date: Optional[date] = None


# ============================================
# AUTHENTICATION SCHEMAS
# ============================================

class AuthItemCreate(BaseModel):
    """Create an auth item."""
    item_type: str  # card, memorabilia, collectible
    inventory_id: Optional[UUID] = None  # For cards
    standalone_item_id: Optional[UUID] = None  # For memorabilia/collectibles
    description: Optional[str] = None
    signer_name: Optional[str] = None
    declared_value: Decimal = Decimal("0")
    fee_per_item: Optional[Decimal] = None


class AuthItemResponse(BaseModel):
    id: UUID
    item_type: str
    inventory_id: Optional[UUID] = None
    standalone_item_id: Optional[UUID] = None
    line_number: Optional[int] = None
    description: Optional[str] = None
    signer_name: Optional[str] = None
    declared_value: Decimal
    fee_per_item: Optional[Decimal] = None
    status: str
    cert_number: Optional[str] = None
    sticker_number: Optional[str] = None
    letter_number: Optional[str] = None
    notes: Optional[str] = None
    
    # Nested details for cards
    player_name: Optional[str] = None
    card_number: Optional[str] = None
    product_line_name: Optional[str] = None
    
    # Nested details for standalone items
    item_name: Optional[str] = None
    item_category: Optional[str] = None
    
    class Config:
        from_attributes = True


class AuthSubmissionCreate(BaseModel):
    """Create a new auth submission."""
    company_id: UUID
    service_level_id: Optional[UUID] = None
    submitter_id: Optional[UUID] = None  # Third-party submitter (NULL = direct)
    date_submitted: date
    items: List[AuthItemCreate]
    submission_number: Optional[str] = None
    reference_number: Optional[str] = None
    shipping_to_cost: Decimal = Decimal("0")
    shipping_to_tracking: Optional[str] = None
    insurance_cost: Decimal = Decimal("0")
    notes: Optional[str] = None


class AuthSubmissionResponse(BaseModel):
    id: UUID
    company_id: UUID
    service_level_id: Optional[UUID] = None
    submitter_id: Optional[UUID] = None
    submission_number: Optional[str] = None
    reference_number: Optional[str] = None
    date_submitted: date
    date_shipped: Optional[date] = None
    date_received: Optional[date] = None
    date_completed: Optional[date] = None
    date_shipped_back: Optional[date] = None
    date_returned: Optional[date] = None
    status: str
    authentication_fee: Decimal
    shipping_to_cost: Decimal
    shipping_to_tracking: Optional[str] = None
    shipping_return_cost: Decimal
    shipping_return_tracking: Optional[str] = None
    insurance_cost: Decimal
    total_items: int
    items_authenticated: int
    total_declared_value: Decimal
    notes: Optional[str] = None
    items: List[AuthItemResponse] = []
    
    # Nested company info
    company_name: Optional[str] = None
    company_code: Optional[str] = None
    service_level_name: Optional[str] = None
    submitter_name: Optional[str] = None
    
    class Config:
        from_attributes = True


class AuthStatusUpdate(BaseModel):
    """Update auth submission status."""
    status: str  # shipped, received, processing, shipped_back, returned
    date_shipped: Optional[date] = None
    date_received: Optional[date] = None
    date_completed: Optional[date] = None
    date_shipped_back: Optional[date] = None
    date_returned: Optional[date] = None
    shipping_to_tracking: Optional[str] = None
    shipping_return_tracking: Optional[str] = None


class AuthResult(BaseModel):
    """Result for a single authenticated item."""
    item_id: UUID
    status: str  # authentic, not_authentic, inconclusive, lost
    cert_number: Optional[str] = None
    sticker_number: Optional[str] = None
    letter_number: Optional[str] = None
    notes: Optional[str] = None


class AuthResultsSubmit(BaseModel):
    """Submit authentication results."""
    item_results: List[AuthResult]
    date_returned: Optional[date] = None
    shipping_return_cost: Decimal = Decimal("0")


class AuthStats(BaseModel):
    """Statistics for authentication."""
    pending_submissions: int
    items_out_for_auth: int
    pending_fees: Decimal
    total_authenticated: int
    pass_rate: float  # Percentage authenticated
    by_item_type: Dict[str, int]  # {"card": 50, "memorabilia": 20, "collectible": 10}
    by_company: Dict[str, int]  # {"PSA-DNA": 40, "JSA": 40}


class AuthPendingByCompany(BaseModel):
    """Pending auth items grouped by company."""
    company_id: UUID
    company_name: str
    company_code: str
    pending_count: int
    pending_value: Decimal
    oldest_submission_date: Optional[date] = None

File: frontend/src/types/grading.ts
/**
 * Grading & Authentication Types
 */

// ============================================
// SHARED TYPES
// ============================================

export interface GradingServiceLevel {
  id: string;
  name: string;
  code: string | null;
  max_value: number | null;
  base_fee: number;
  estimated_days: number | null;
  is_active: boolean;
}

export interface GradingCompanyWithLevels {
  id: string;
  name: string;
  code: string;
  website: string | null;
  service_type: 'grading' | 'authentication' | 'both';
  is_active: boolean;
  service_levels: GradingServiceLevel[];
}

export interface PendingByCompany {
  company_id: string;
  company_name: string;
  company_code: string;
  pending_count: number;
  pending_value: number;
  oldest_submission_date: string | null;
}

// ============================================
// CARD GRADING TYPES
// ============================================

export interface CardGradingItem {
  id: string;
  inventory_id: string | null;
  checklist_id: string | null;
  line_number: number | null;
  declared_value: number;
  fee_per_card: number | null;
  was_signed: boolean;
  status: 'pending' | 'graded' | 'authentic' | 'altered' | 'counterfeit' | 'ungradeable' | 'lost';
  grade_value: number | null;
  auto_grade: number | null;
  cert_number: string | null;
  label_type: string | null;
  notes: string | null;
  // Nested details
  player_name?: string;
  card_number?: string;
  product_line_name?: string;
}

export interface CardGradingSubmission {
  id: string;
  company_id: string;
  service_level_id: string | null;
  submission_number: string | null;
  reference_number: string | null;
  date_submitted: string;
  date_shipped: string | null;
  date_received: string | null;
  date_graded: string | null;
  date_shipped_back: string | null;
  date_returned: string | null;
  status: 'pending' | 'shipped' | 'received' | 'grading' | 'shipped_back' | 'returned' | 'cancelled';
  grading_fee: number;
  shipping_to_cost: number;
  shipping_to_tracking: string | null;
  shipping_return_cost: number;
  shipping_return_tracking: string | null;
  insurance_cost: number;
  total_cards: number;
  cards_graded: number;
  total_declared_value: number;
  notes: string | null;
  items: CardGradingItem[];
  // Nested company info
  company_name?: string;
  company_code?: string;
  service_level_name?: string;
}

export interface CardGradingItemCreate {
  inventory_id: string;
  checklist_id?: string;
  declared_value?: number;
  fee_per_card?: number;
  was_signed?: boolean;
}

export interface CardGradingSubmissionCreate {
  company_id: string;
  service_level_id?: string;
  date_submitted: string;
  items: CardGradingItemCreate[];
  submission_number?: string;
  reference_number?: string;
  shipping_to_cost?: number;
  shipping_to_tracking?: string;
  insurance_cost?: number;
  notes?: string;
}

export interface CardGradingStatusUpdate {
  status: string;
  date_shipped?: string;
  date_received?: string;
  date_graded?: string;
  date_shipped_back?: string;
  date_returned?: string;
  shipping_to_tracking?: string;
  shipping_return_tracking?: string;
}

export interface CardGradeResult {
  item_id: string;
  status: string;
  grade_value?: number;
  auto_grade?: number;
  cert_number?: string;
  label_type?: string;
  notes?: string;
}

export interface CardGradingResultsSubmit {
  item_results: CardGradeResult[];
  date_returned?: string;
  shipping_return_cost?: number;
}

export interface CardGradingStats {
  pending_submissions: number;
  cards_out_for_grading: number;
  pending_fees: number;
  total_graded: number;
  grade_distribution: Record<string, number>;
  gem_rate: number;
  avg_grade?: number;
  by_company?: Record<string, number>;
}

// ============================================
// AUTHENTICATION TYPES
// ============================================

export type AuthItemType = 'card' | 'memorabilia' | 'collectible';

export interface AuthItem {
  id: string;
  item_type: AuthItemType;
  inventory_id: string | null;
  standalone_item_id: string | null;
  line_number: number | null;
  description: string | null;
  signer_name: string | null;
  declared_value: number;
  fee_per_item: number | null;
  status: 'pending' | 'authentic' | 'not_authentic' | 'inconclusive' | 'lost';
  cert_number: string | null;
  sticker_number: string | null;
  letter_number: string | null;
  notes: string | null;
  // Nested card details
  player_name?: string;
  card_number?: string;
  product_line_name?: string;
  // Nested standalone details
  item_name?: string;
  item_category?: string;
}

export interface AuthSubmission {
  id: string;
  company_id: string;
  service_level_id: string | null;
  submission_number: string | null;
  reference_number: string | null;
  date_submitted: string;
  date_shipped: string | null;
  date_received: string | null;
  date_completed: string | null;
  date_shipped_back: string | null;
  date_returned: string | null;
  status: 'pending' | 'shipped' | 'received' | 'processing' | 'shipped_back' | 'returned' | 'cancelled';
  authentication_fee: number;
  shipping_to_cost: number;
  shipping_to_tracking: string | null;
  shipping_return_cost: number;
  shipping_return_tracking: string | null;
  insurance_cost: number;
  total_items: number;
  items_authenticated: number;
  total_declared_value: number;
  notes: string | null;
  items: AuthItem[];
  // Nested company info
  company_name?: string;
  company_code?: string;
  service_level_name?: string;
}

export interface AuthItemCreate {
  item_type: AuthItemType;
  inventory_id?: string;
  standalone_item_id?: string;
  description?: string;
  signer_name?: string;
  declared_value?: number;
  fee_per_item?: number;
}

export interface AuthSubmissionCreate {
  company_id: string;
  service_level_id?: string;
  date_submitted: string;
  items: AuthItemCreate[];
  submission_number?: string;
  reference_number?: string;
  shipping_to_cost?: number;
  shipping_to_tracking?: string;
  insurance_cost?: number;
  notes?: string;
}

export interface AuthStatusUpdate {
  status: string;
  date_shipped?: string;
  date_received?: string;
  date_completed?: string;
  date_shipped_back?: string;
  date_returned?: string;
  shipping_to_tracking?: string;
  shipping_return_tracking?: string;
}

export interface AuthResult {
  item_id: string;
  status: string;
  cert_number?: string;
  sticker_number?: string;
  letter_number?: string;
  notes?: string;
}

export interface AuthResultsSubmit {
  item_results: AuthResult[];
  date_returned?: string;
  shipping_return_cost?: number;
}

export interface AuthStats {
  pending_submissions: number;
  items_out_for_auth: number;
  pending_fees: number;
  total_authenticated: number;
  pass_rate: number;
  by_item_type: Record<string, number>;
  by_company: Record<string, number>;
}

// ============================================
// LEGACY ALIASES (for backwards compatibility)
// ============================================

/** @deprecated Use GradingServiceLevel */
export type ServiceLevel = GradingServiceLevel;

/** @deprecated Use GradingCompanyWithLevels */
export type GradingCompany = GradingCompanyWithLevels;